### DISCUSSION TITLE
How does your Gif / Video making coming through?

### DISCUSSION THREAD

**Discussion by Vishal-Upendran** [Vishal-Upendran/IUCAASummerSchool2021#4] [Show and tell]
I had asked you folks to make a gif/video of the solar dataset showed during the tutorial. How is it going?

**Comment by AnushkaTilekar** [comment]
```
from astropy.io import fits

mkdir 
Data/

wget: 
https://www.lmsal.com/solarsoft/irisa/data/level2_compressed/2014/10/22/20141022_081850_3860261381/iris_l2_20141022_081850_3860261381_SDO.tar.gz

tar -xvf iris_l2_20141022_081850_3860261381_SDO.tar.gz

aia = fits.open("aia_l2_20141022_081850_3860261381_171.fits")
print(aia.info())

aia[0].header
aiadata = aia[0].data
aiahead = aia[0].header
print(aiadata.shape)

from astropy.wcs import WCS
wcs  = WCS(aiahead)
print(wcs)
#Plot at first time step.
ind = 50
plt.figure(figsize=(8,8))
plt.subplot(projection = wcs.dropaxis(-1))
plt.imshow(np.log10(aiadata[ind]),cmap='Blues_r')


from celluloid import Camera
# Write code to load data, take frame and then make video.
# Check the example code here and adapt it: https://github.com/jwkvam/celluloid.
fig = plt.figure(figsize=(8,8))
camera = Camera(fig)
for ind in np.arange(10):
    val = np.log10(aiadata[ind])
    plt.imshow(val,cmap='Blues_r',clim=(np.nanpercentile(np.log10(aiadata),5),np.nanpercentile(np.log10(aiadata),99)),origin='lower')
    camera.snap()
from IPython.display import HTML
animation = camera.animate()
HTML(animation.to_html5_video())



```


Tried to copy-paste this code on my Visual Studio code. Yes I did write the commands to download/import these files in terminal  but the errors are still persisiting.
Getting these errors := 

Import "astropy.io" could not be resolved
"mkdir" is not defined
"Data" is not defined
"tar" is not defined
"xvf" is not defined
"iris_l2_20141022_081850_3860261381_SDO" is not defined
Import "astropy.wcs" could not be resolved
"plt" is not defined
"np" is not defined
Import "celluloid" could not be resolved
Import "IPython.display" could not be resolved




.

**Reply by Vishal-Upendran** [reply]
That wont work. Click on terminal of VSCode. On the terminal do:

`wget:  https://www.lmsal.com/solarsoft/irisa/data/level2_compressed/2014/10/22/20141022_081850_3860261381/iris_l2_20141022_081850_3860261381_SDO.tar.gz`

`tar -xvf iris_l2_20141022_081850_3860261381_SDO.tar.gz`

`pip install celluloid`.

And remove the first 2 lines, `mkdir Data/` from the python file. Also remove the HTML() function, it is used only in Jupyter notebook. Check here for reference: https://github.com/jwkvam/celluloid

Basically, your code should be:

``from astropy.io import fits
aia = fits.open("aia_l2_20141022_081850_3860261381_171.fits")
print(aia.info())

aia[0].header
aiadata = aia[0].data
aiahead = aia[0].header
print(aiadata.shape)

from astropy.wcs import WCS
wcs  = WCS(aiahead)
print(wcs)
#Plot at first time step.
ind = 50
plt.figure(figsize=(8,8))
plt.subplot(projection = wcs.dropaxis(-1))
plt.imshow(np.log10(aiadata[ind]),cmap='Blues_r')

from celluloid import Camera
fig = plt.figure(figsize=(8,8))
camera = Camera(fig)
for ind in np.arange(10):
    val = np.log10(aiadata[ind])
    plt.imshow(val,cmap='Blues_r',clim=(np.nanpercentile(np.log10(aiadata),5),np.nanpercentile(np.log10(aiadata),99)),origin='lower')
    camera.snap()
animation = camera.animate()
animation.save('animation.mp4')``

Does it work now? If you don't install the packages from terminal in the first 2 lines, it wont work...

**Reply by yuvrajmuralichandran** [reply]
The line 2 of the code:

`aia = fits.open("aia_l2_20141022_081850_3860261381_171.fits")`

returns the following error:

`FileNotFoundError: [Errno 2] No such file or directory: 'aia_l2_20141022_081850_3860261381_171.fits`

Should we download the file before executing the code or is there a way to access the FITS directly from the website?

**Reply by Vishal-Upendran** [reply]
Yes. Do the `wget` and `tar` first, on a terminal, to download the file and then to extract it out. Then you can open the `aia_` file. 

If you are using colab, run the cell with shift+enter, it will download the files.

**Reply by yuvrajmuralichandran** [reply]
I am trying it in Spyder (Anaconda) IDE, is it same procedure?

I tried in the IDLE of Python this:
`>>>wget:https://www.lmsal.com/solarsoft/irisa/data/level2_compressed/2014/10/22/20141022_081850_3860261381/iris_l2_20141022_081850_3860261381_SDO.tar.gz`

but it shows:
`SyntaxError: invalid syntax`

**Reply by Vishal-Upendran** [reply]
No no, you should use terminal. That's a bash command. If that doesn't work, just go the link `https://www.lmsal.com/solarsoft/irisa/data/level2_compressed/2014/10/22/20141022_081850_3860261381/iris_l2_20141022_081850_3860261381_SDO.tar.gz`. It will ask you to download the data. 

Save it in the same folder as you code and extract it. That should work.

**Reply by AnushkaTilekar** [reply]
After typing this in the terminal =
`wget: https://www.lmsal.com/solarsoft/irisa/data/level2_compressed/2014/10/22/20141022_081850_3860261381/iris_l2_20141022_081850_3860261381_SDO.tar.gz`




It gives the the following error =

wget: : The term 'wget:' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, 
verify that the path is correct and try again.
At line:1 char:1
+ wget: https://www.lmsal.com/solarsoft/irisa/data/level2_compressed/20 ...
+ ~~~~~
    + CategoryInfo          : ObjectNotFound: (wget::String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException

**Comment by Vishal-Upendran** [comment]
You shouldn't have a : after wget. The wget command is

wget file_url

**Reply by AnushkaTilekar** [reply]
oh okk. 
I put it in accordingly and now it is 'reading the bytes' =

''Reading web response
    Reading response stream... (Number of bytes read: 1999041918)
''


and that no. is continuously increasing. But it is going on for 20-25 mins now. How much time does it take to complete? 
And As it is taking so much time to complete, does it mean that the file size is too large?.. I mean, should I be worried about the storage size availability on my laptop ?

**Comment by Vishal-Upendran** [comment]
The data is 5 GB, it will take some time depending on your internet speed!

That's why I was asking you to run it on colab, since colab will download
using net speed of Google servers, and not your internet.

After it's downloaded,  extract it with tar, and then run python commands.
It should work.

Did you also install celluloid using pip on the terminal?

**Reply by AnushkaTilekar** [reply]
yes I did install celluloid on terminal

**Reply by Vishal-Upendran** [reply]
Excellent. How did it go?

**Reply by AnushkaTilekar** [reply]
I was facing dificulty on loading the data on VS Code. So I shifted to Colab. I wanted a multicolor gif, so I wrote these blocks of codes =
But it is giving a stationary image instead of gif.

----------------------------------------------------------------------------

# CODE_BLOCK-1 

!pip install celluloid


-------------------------------------------------------------------------

# CODE_BLOCK-2

!wget https://www.lmsal.com/solarsoft/irisa/data/level2_compressed/2014/10/22/20141022_081850_3860261381/iris_l2_20141022_081850_3860261381_SDO.tar.gz

----------------------------------------------------------------------------------------------

# CODE_BLOCK-3

!tar -xvf iris_l2_20141022_081850_3860261381_SDO.tar.gz


--------------------------------------------------------------------------------------------------

# CODE_BLOCK-4

import numpy as np
from matplotlib import pyplot as plt

-------------------------------------------------------------------------------------------------

# CODE_BLOCK-5

from astropy.io import fits

----------------------------------------------------------------------------------------------

# CODE_BLOCK-6

from astropy.wcs import WCS

-------------------------------------------------------------------------------------------

# CODE_BLOCK-7

from celluloid import Camera

------------------------------------------------------------------------------------------------

# CODE_BLOCK-8

from IPython.display import HTML

---------------------------------------------------------------------------------------------------

# CODE_BLOCK-9

fig = plt.figure(figsize=(8,8))

---------------------------------------------------------------------------------------------------

# CODE_BLOCK-10

aia = fits.open("aia_l2_20141022_081850_3860261381_171.fits")
print(aia.info())

--------------------------------------------------------------------------------------------------

# CODE_BLOCK-11

aiadata = aia[0].data
aiahead = aia[0].header
camera = Camera(fig)

--------------------------------------------------------------------------------------------------

# CODE_BLOCK-12  (everything below is in this single block)

for ind in np.arange(8):
    val = np.log10(aiadata[ind])
    plt.imshow(val,cmap='Reds_r',clim=(np.nanpercentile(np.log10(aiadata),20),np.nanpercentile(np.log10(aiadata),99)),origin='lower')
    camera.snap()
    animation = camera.animate()
    HTML(animation.to_html5_video())


    val = np.log10(aiadata[ind])
    plt.imshow(val,cmap='Blues_r',clim=(np.nanpercentile(np.log10(aiadata),60),np.nanpercentile(np.log10(aiadata),99)),origin='lower',alpha=0.5)
    camera.snap()
    animation = camera.animate()
    HTML(animation.to_html5_video())


    val = np.log10(aiadata[ind])
    plt.imshow(val,cmap='Greens_r',clim=(np.nanpercentile(np.log10(aiadata),60),np.nanpercentile(np.log10(aiadata),99)),origin='lower',alpha=0.5)
    camera.snap()
    animation = camera.animate()
    HTML(animation.to_html5_video())
`
(#Note == all these 3 sub-blocks from 'val' to 'HTML' after the 'for', are written under 'for' with all 3 indented to right side. This indentation might not be visible the comment.)

---------------------------------------------------------------------------------------------------------

**Reply by Vishal-Upendran** [reply]
Okay so there are multiple issues here.

1. **Way colours work** : A single channel image like the one here gives you only intensities, which can be alloted to any colourmap like Blues, Greens, etc. If you want to make the 3-color image, you need 3 intensities, one each of Red, Blue and Green. 
2. **Way camera works**: Think of the figure as a canvas, on which you can make multiple drawings. Each `imshow()` is like a drawing on a canvas. 
   2.1  When you first do `Camera(fig)`, you are saying "let's look at this canvas called fig".
   2.2. Then you say `imshow()` , which means you are drawing the image. 
   2.3. Finally, you do `camera.snap()`, meaning you take a snapshot of the image, and repeat the process. 
   2.4. After all images are drawn and the snapshots taken, you say `camera.animate()`, meaning you want to animate the bunch of snapshots. 

In your code, you are plotting the same image, on the same canvas, with different colourmaps. The reason you don't get an animation is because the `animate()` and `HTML(...)` commands should be outside the for loop, possibly in a different cell. 

Check the tutorial for reference on 3-channel images: https://vishal-upendran.github.io/random/2021/06/01/python_intro.html. You can click on the `open in colab` to directly open in colab.

**Reply by AnushkaTilekar** [reply]
Ohk. I understood the explanation.👍
I tried it by shifting those lines to separate cell, but then I'm getting this blank gif......

--------------------------------------------------------------------------------

'#code block 1


camera = Camera(fig)

for ind in np.arange(8):
    val = np.log10(aiadata[ind])
    plt.imshow(val,cmap='Reds_r',clim=(np.nanpercentile(np.log10(aiadata),20),np.nanpercentile(np.log10(aiadata),99)),origin='lower')
    camera.snap()


    val = np.log10(aiadata[ind])
    plt.imshow(val,cmap='Blues_r',clim=(np.nanpercentile(np.log10(aiadata),60),np.nanpercentile(np.log10(aiadata),99)),origin='lower',alpha=0.5)
    camera.snap()


    val = np.log10(aiadata[ind])
    plt.imshow(val,cmap='Greens_r',clim=(np.nanpercentile(np.log10(aiadata),60),np.nanpercentile(np.log10(aiadata),99)),origin='lower',alpha=0.5)
    camera.snap()

-----------------------------------------------------------------------------------------------------------------
#code block 2

animation = camera.animate()
HTML(animation.to_html5_video())`
---------------------------------------------------------------------------------------------------------------------------



https://user-images.githubusercontent.com/82459979/121809060-f1373100-cc78-11eb-84a4-883e00715cfe.mp4

**Reply by Vishal-Upendran** [reply]
In your `imshow()`, remove the `clim=()` bit from all 3 lines. Also, do `val = np.log10(aiadata[ind])` only once (before Reds_r) and see if it works.

**Reply by AnushkaTilekar** [reply]
no it is still not working..

**Reply by Vishal-Upendran** [reply]
Is this your code?

```python
fig=plt.figure()
camera=Camera(fig)
for ind in np.arange(8):
    val = np.log10(aiadata[ind])
    plt.imshow(val,cmap='Blues_r',origin='lower',alpha=0.5)
    camera.snap()
    plt.imshow(val,cmap='Greens_r',origin='lower',alpha=0.5)
    camera.snap()```

# and then you have

animation = camera.animate()
HTML(animation.to_html5_video())

**Reply by AnushkaTilekar** [reply]
yes.
This is the first line
fig = plt.figure(figsize=(8,8))

and before the line of blue, there are these lines 
    plt.imshow(val,cmap='Reds_r',origin='lower')
    camera.snap()

**Reply by Vishal-Upendran** [reply]
Hmm... What do you get when you do `print(val)` ? Is it some proper values or bunch of nans?

Try this in first line: `val = np.log10(np.copy(aiadata[ind]))`. Does it work? You are getting white colour because most of the image is nan, for some reason.

And try restarting your notebook and run it. Maybe there are some saved values causing error.

**Reply by AnushkaTilekar** [reply]
It is giving this output -
`[[2.8591383 2.822168  2.7923918 ... 2.049218  2.1003704 2.146128 ]
 [2.924796  2.8808136 2.8438554 ... 2.0170333 2.0374265 2.0606978]
 [2.9894497 2.94939   2.9138138 ... 1.9542425 1.94939   1.9542425]
 ...
 [2.2479732 2.2624512 2.2787535 ... 3.3436055 3.3434086 3.3416324]
 [2.2479732 2.252853  2.2600713 ... 3.3539162 3.3694015 3.376577 ]
 [2.2479732 2.2479732 2.2552724 ... 3.3558345 3.3743818 3.3836358]]`


and now I'm getting a black coloured gif frame square output with No gif in it....!

I've also tried starting the notebook several times and everytime this blank output......

**Reply by Vishal-Upendran** [reply]
Hmm. This is so weird, since there are no Nans in the data. 

What do you get if you do plot only a single colour? Do you get the gif displayed, or the animation saved as an mp4? 

```python
aia = fits.open("aia_l2_20141022_081850_3860261381_171.fits")
aiadata = aia[0].data
fig=plt.figure(figsize=(8,8))
camera=Camera(fig)
for ind in np.arange(8):
    val = np.log10(aiadata[ind])
    plt.imshow(val,cmap='Blues_r',origin='lower')
    camera.snap()

# and then you have

animation = camera.animate()
animation.save('animation.mp4')
HTML(animation.to_html5_video())```

**Reply by AnushkaTilekar** [reply]
yes, it works perfectly for single colours. 
For this - 
`fig = plt.figure(figsize=(8,8))
camera = Camera(fig)

for ind in np.arange(200):
    val = np.log10(aiadata[ind])
   
    plt.imshow(val,cmap='Reds_r',origin='lower')
    camera.snap()
`

we get the output of this stationary image -

![456](https://user-images.githubusercontent.com/82459979/122191723-9be56480-ceb0-11eb-8399-ae26d32a3c43.png)


and for this -

`from IPython.display import HTML

animation = camera.animate()

animation.save('animation.mp4')

HTML(animation.to_html5_video())`





we get this gif - 

https://user-images.githubusercontent.com/82459979/122191876-bcadba00-ceb0-11eb-9796-5a84f76d8ce3.mp4




I just don't understand why it's not working for multiple colours ? or that how to make it work for multiple colours ?..

**Reply by Vishal-Upendran** [reply]
Interesting. This works perfectly for me - could you check in your notebook?

```python
fig = plt.figure(figsize=(8,8))
camera = Camera(fig)
for ind in np.arange(10):
    val = np.log10(aiadata[ind])
    plt.imshow(val,cmap='Blues_r',origin='lower')
    camera.snap()
    plt.imshow(val,cmap='Reds_r',origin='lower')
    camera.snap()
    plt.imshow(val,cmap='Greens_r',origin='lower')
    camera.snap()
from IPython.display import HTML
animation = camera.animate()
HTML(animation.to_html5_video())
```

**Reply by AnushkaTilekar** [reply]
Yes. It worked ! And I even tried to take this learning one step further. I tried to make a complete videos instead of only gifs and I was able to make videos as up to as 10 minutes from the single .fits file 😄 !!!! I also tried to add dynamic labels as well.
Can you please look at them and tell me how they are ? 
https://drive.google.com/drive/folders/1JoMsN9pRQs3JmglS7gYSbY9rzCTu9ZHF?usp=sharing


Although, I wonder that when we downloaded the .fits images, there are total 9 .fits images. But when I opened ech of the image I found that the range of index values is different for each of these images :-

 ind range for image 1 is = (-3008 numbers)--(0)--(+3007 numbers) .

 ind range for image 2 is = (-3006 numbers)--(0)--(+3005 numbers) .

 ind range for image 3 is = (-3010 numbers)--(0)--(+3009 numbers) .

 ind range for image 4 is = (-3010 numbers)--(0)--(+3009 numbers) .

 ind range for image 5 is = (-3010 numbers)--(0)--(+3009 numbers) .

 ind range for image 6 is = (-3007 numbers)--(0)--(+3006 numbers) .

 ind range for image 7 is = (-1515 numbers)--(0)--(+1514 numbers) .

 ind range for image 8 is = (-1515 numbers)--(0)--(+1514 numbers) .

 ind range for image 9 is = (-1525 numbers)--(0)--(+1524 numbers) .


.
I wonder why is that so though ?.......




.

**Reply by Vishal-Upendran** [reply]
Wonderful! These look very good!

Very good observation on the different ranges. This comes due to two things:

1. Difference between 3000 and 1500 indices: This occurs due to the 7,8,9 being available every 24 seconds or so, while the others are available every 12 seconds. You should check the `header` of the fits file - just print it out. Check for values like `CDELT1`, `CDELT2` etc. This gives you the all dx, dy and dt of the data. To find the delta corresponding to time, check `CTYPE`, and see which has info on what that value is (time? latitude? longitude? etc.). Compare this for the files and see!
2. Variation about 3000 and 1500: The cadence, or data availability is approximately 12 or 24 seconds, so sometimes data might be missing, causing changes in the index range.

Good work!

**Reply by AnushkaTilekar** [reply]
Thank you ! 😄.
Yes, I checked the values, there is variation in them as you mentioned. 👍 Now I understood why the range is different.
I also tried to obtain a gif/video of this image, but I'm only getting a video with static image... The index value keeps changing but there is no visible movement in the image..

https://user-images.githubusercontent.com/82459979/123423744-90dfb200-d5dd-11eb-850c-5c26d7b1d46e.mp4

This is the code that I was trying to do it with. I also tried to do what I did it for previous fits files, but didn't work.......

https://colab.research.google.com/drive/10nndH9nzD0M5radH09D0OEm5_jbE1VKW?usp=sharing

**Reply by Vishal-Upendran** [reply]
That's because you are not plotting the correct variable. Check your code:, you are doing:
```python
plt.imshow(tmp_aia, origin='lower', interpolation='nearest')```

while it should be :
```python
plt.imshow(val_tmp_aia, origin='lower', interpolation='nearest')```

Also, in your cell which does `for index in range (-1515,1514,1):`, you don't need to use the for loop at all. Check numpy documentation for `axis` parameter, and vectorize your code.

**Comment by baibhavsrivastava** [comment]
This is what I could come up with, but the notebook showed some runtime warning about some numbers having a zero in the denominator and invalid numbers in log, no idea where did that come from. But even with this, the plot turned out to be quite fine, not sure how, I was expecting just NaNs after I saw the warning.

https://user-images.githubusercontent.com/66297621/120846695-20baaf00-c590-11eb-8c96-22f546c55ed2.mp4

**Reply by Vishal-Upendran** [reply]
That's awesome! 

Nans: They come because certain pixels in the images have negative values --  these are the unphysical intensities I talked about. What do you get when you take log of negative values?

Also, in most colorscales in matplotlib, Nan values are given a value 0 (or black) or 1 (white).  So you wouldn't see then so conspicuously in the data. 

Well done! If you see the video, you get bursts of intensities near the center - that is because magnetic field lines are cut and then rejoin. With the video, one can estimate the time scales such bursts last, and hence know how much energy is dumped over a give time in the corona!

**Reply by baibhavsrivastava** [reply]
Now I understood why the NaNs and why can't I see them in the plot, thanks a lot!
It's amazing how one can extract so much information from these data. I also noticed a kind of overall dimming in the video at around 10s, was it because of some instrumental errors?
I will also try to recreate this video with multiple filters...

**Reply by Vishal-Upendran** [reply]
Good observation.

For plotting each frame, matplotlib selects max and min intensities in the image and uses them. At 10 sec, near (100,300), the pixel seems to become vastly more brighter than the remaining field of view. This causes everything else to become darker. It is an artefact of not setting your colour limits. 

You can prevent this from happening by setting clim=(min_value,max_value). In that case, all frames will have the same max and min intensities -- any values outside this limit will be set to the limits.

**Reply by baibhavsrivastava** [reply]
Will try that, thanks a lot!
